### PR TITLE
Support trace_subgraph in _MakefxTracer

### PR DESCRIPTION
**Pull request by ydwu4**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125363
* #125267

Adds trace_subgraph to _MakefxTracer, the motivation is in https://github.com/pytorch/pytorch/pull/122972. Also migrate all existing usage of reenter_make_fx to the new sub-tracer. Previously, the torch function mode for creating torch_fn metadata won't be re-enetered when we're in ProxyTensorMode (since it's inside of __torch_function__). This PR reconstruct the torch function mode based on parent tracer's config and reentered the torch function mode so the metadata is shown in the graph.

**Test Plan:**
Existing tests. We have a bunch of make_fx tests for cond, map and while_loop. Also remove expected failure for torch_fn since reenter_make_fx is able to re-construct torch function modes. 

Also fixes https://github.com/pytorch/pytorch/issues/124643

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/fx/experimental/proxy_tensor.py [hunk 1]
```diff
@@ -1224,32 +1245,62 @@ def _init_modes_from_inputs(self, f, args):
...
+            self.fake_tensor_mode = parent_tracer.fake_tensor_mode
+
+            def _create_sub_fx_tracer(parent_tracer):
+                if type(parent_tracer) == PythonKeyTracer:
+                    sub_tracer = PythonKeyTracer()
+                elif type(parent_tracer) == _ModuleStackTracer:
+                    sub_tracer = _ModuleStackTracer(parent_tracer.scope_root)
+                    # We want sub_tracer to share the same states as parent tracer
+                    # for below states to correctly compute some metadata
+                    sub_tracer.scope_root = parent_tracer.scope_root
```
Comment (ydwu4) [inline torch/fx/experimental/proxy_tensor.py:1291]: Add a test for `nn_module_stack`

Comment (ydwu4) [reply torch/fx/experimental/proxy_tensor.py]: Its a bit difficult to test in this diff: control flow operators turn on torch.compile by default so at the time we're doing reenter_make_fx in Proxy mode, the subgraphs have been flattened. and we lose track of these metadata anyway. This nees a bit more design of keeping original user's metadata when we turn on torch.compile by default.

File: torch/fx/experimental/proxy_tensor.py [hunk 2]
```diff
@@ -1166,25 +1168,45 @@ def __init__(
         self._error_on_data_dependent_ops: bool = _error_on_data_dependent_ops
 
         # All context managers and their states should be initialized before tracing based on the inputs
-        # and configurations.
-        # After tracing, their states should be cleaned except for shape_env.
-        null_ctx_type = type(nullcontext)
+        # and configurations. After tracing, their states should be cleaned except for shape_env.
+        # Rememer to specify how to intialize it from user inputs and from parent tracer whenever
```
Comment (Chillee) [inline torch/fx/experimental/proxy_tensor.py:1172]: spelling "Rememer"

**Review by Chillee — COMMENTED** [review]
Please also modify FlexAttention to use this tracer.

**Review by Chillee — APPROVED** [review]

**Pull request closed without merging**